sourced/ml/repo2/token_map.py

```diff
@@ -0,0 +1,29 @@
+from operator import add
+
+from sourced.ml.engine import get_tokens
+from sourced.ml.repo2.base import Transformer
+from sourced.ml.token_parser import TokenParser
+
+
+class TokenMapTransformer(Transformer):
+    def __init__(self,  token_parser, prune_size=1, **kwargs):
+        super().__init__(**kwargs)
+        self.token_parser = token_parser
+        self.prune_size = prune_size
+
+    def __call__(self, uasts):
+        """
+        Make tokens list and token2index mapping from provided uasts.
+        token2index is broadcasted dictionary to use it in workers. Maps token to its index.
+        """
+        tokens = get_tokens(uasts).rdd \
+            .flatMap(lambda r: [(t, 1) for token in r.tokens for t in self.token_parser(token)]) \
```

Comment:
Can be optimized with `countByKey()`

Reply:
depends from point of view. 
it returns python dict, and then we do pruning. So, it is much better to use this approach for master node memory since we have a huge number of rare tokens. 

Prefer to keep as is.

Reply:
You did not get me. `reduceByKey(add)` is exactly the same as `countByKey()` but slower. Everything else is left untouched.

Reply:
I get what you mean, but no, it is not the same.

if I run 
```python
self.get_tokens(uasts).rdd \
            .flatMap(lambda r: [(t, 1) for token in r.tokens for t in self.token_parser(token)]) \
            .reduceByKey(add)
```
I will get the `rdd`, nothing is executed.

If I run 
```python
self.get_tokens(uasts).rdd \
            .flatMap(lambda r: [(t, 1) for token in r.tokens for t in self.token_parser(token)]) \
            .countByKey()
```
I will get python `dict`. I Do not know why it has such behavior, but I don't what to have dict here.

Just now I check it one time more.

Reply:
Yes, you are right. From official docs:

```
Count the number of elements for each key, and return the result to the master as a dictionary.
```

Indeed, if you need an rdd, this does not work. Sorry for misleading.